client/internal/retry/do.go

```diff
@@ -83,62 +93,59 @@ func NewOptions(logger *log.Logger, timeout time.Duration, delay time.Duration,
 }
 
 // Do run retry function until response of request satisfy check function, or ends early according to configuration.
-func Do(retryFunc Func, opt Options) error {
-
-	startTime := time.Now()
-	endTime := startTime.Add(opt.Timeout)
-	timeout := func() bool {
-		return time.Now().After(endTime)
-	}
-	willTimeoutAfterDelay := func() bool {
-		return time.Now().Add(opt.Delay).After(endTime)
-	}
-	accumulatedErrs := make([]error, 0, opt.Retries+1)
-
-	// check if it is already timeout
-	if timeout() {
-		return fmt.Errorf("retry func timeout before any attempt")
+func Do(ctx context.Context, retryFunc Func, opt Options) error {
+	// set up context
+	ctxToUse := ctx
+	if opt.Timeout > 0 {
+		var cancel context.CancelFunc
+		ctxToUse, cancel = context.WithTimeout(ctx, opt.Timeout)
+		defer cancel()
 	}
+	return doInternal(ctxToUse, retryFunc, opt)
+}
 
-	// attempt once before starting retries
-	ok, err := retryFunc()
-	accumulatedErrs = append(accumulatedErrs, err)
-	if err != nil && opt.EarlyExitOnError {
-		return fmt.Errorf("error in retry func, cause=%w", err)
+func willTimeoutAfterDelay(ctx context.Context, delay time.Duration) bool {
+	ddl, ok := ctx.Deadline()
+	if !ok {
+		return false // no deadline is set, so we will never time out after delay
 	}
-	if ok {
-		return nil
-	}
-
-	// retry starts
-	for retries := 1; opt.Retries < 0 || retries <= opt.Retries; retries++ {
-
-		// if timeout now or will time out during delay between requests
-		if timeout() || willTimeoutAfterDelay() {
-			return fmt.Errorf("timeout at attempt %d/%d, after %s", retries, opt.Retries, time.Now().Sub(startTime))
-		}
-		time.Sleep(opt.Delay)
-
-		opt.Logger.Printf("[RETRY] attempt=%d/%d\n", retries, opt.Retries)
+	return time.Now().Add(goutil.Max(delay, 0)).After(ddl)
+}
 
-		// attempt
-		ok, err = retryFunc()
-		accumulatedErrs = append(accumulatedErrs, err)
-		if err != nil && opt.EarlyExitOnError {
-			return fmt.Errorf("error in retry func, cause=%w", err)
-		}
-		if ok {
-			opt.Logger.Println("[RETRY] success")
-			return nil
+func doInternal(ctx context.Context, retryFunc Func, opt Options) error {
+	// setup time
+	startTime := time.Now()
+	// setup errors
+	// retryErrors[i] = nil: no error occur at this attempt
+	// retryErrors[i] != nil: error occur at this attempt
+	retryErrors := make([]error, goutil.Max(opt.Retries, 0)+1)
+
+	for attempt := 0; attempt <= opt.Retries || opt.Retries < 0; attempt++ {
+		select {
+		case <-ctx.Done():
+			// context timeout/cancelled
+			return newTimeoutErrorf("%s at attempt=%d/%d, after=%s, errors:\n%w\n", ctx.Err(), attempt, opt.Retries, time.Since(startTime), errors.Join(retryErrors...))
```

Comment:
here in case of `ctx.Done()`, which means the context is cancelled or timeout, we return directly with an error, without further retry attempt